### PR TITLE
Publish function metrics for linux containers

### PR DESCRIPTION
**Pull request by divyagandhii**

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Diagnostics/LinuxContainerEventGenerator.cs [hunk 1]
```diff
@@ -24,6 +28,7 @@ public LinuxContainerEventGenerator(IEnvironment environment, Action<string> wri
                 _consoleEnabled = false;
             }
             _containerName = _environment.GetEnvironmentVariable(EnvironmentSettingNames.ContainerName)?.ToUpperInvariant();
+            _metricsPublisher = new LinuxContainerMetricsPublisher(environment, standbyOptions, LogMetricsPublishEvent, httpClient);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Diagnostics/LinuxContainerEventGenerator.cs]: Instead of passing the LogMetricsPublishEvent func like this, if you instead just have LinuxContainerMetricsPublisher take an ILogger<LinuxContainerMetricsPublisher> ctor param via DI, you can remove this entire line and get the LinuxContainerMetricsPublisher instance via DI. That would be preferable.

File: src/WebJobs.Script.WebHost/Diagnostics/LinuxContainerEventGenerator.cs [hunk 2]
```diff
@@ -71,7 +76,7 @@ public override void LogFunctionMetricEvent(string subscriptionId, string appNam
         {
             string hostVersion = ScriptHost.Version;
 
-            _writeEvent($"{ScriptConstants.LinuxMetricEventStreamName} {subscriptionId},{appName},{functionName},{eventName},{average},{minimum},{maximum},{count},{hostVersion},{eventTimestamp.ToString(EventTimestampFormat)},{NormalizeString(data)},{_containerName},{StampName},{TenantId},,,,");
+            _writeEvent($"{ScriptConstants.LinuxMetricEventStreamName} {subscriptionId},{appName},{functionName},{eventName},{average},{minimum},{maximum},{count},{hostVersion},{eventTimestamp.ToString(EventTimestampFormat)},{NormalizeString(data)},{_containerName},{StampName},{TenantId}");
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Diagnostics/LinuxContainerEventGenerator.cs]: Why the format change?

File: src/WebJobs.Script.WebHost/Metrics/IMetricsPublisher.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Metrics
+{
+    public interface IMetricsPublisher
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/IMetricsPublisher.cs]: Where is this interface used?

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Metrics
+{
+    public class LinuxContainerMetricsPublisher
+    {
+        public const string PublishMemoryActivityPath = "/memoryactivity";
+        public const string PublishFunctionActivityPath = "/functionactivity";
+
+        public const string ContainerNameHeader = "x-ms-functions-container-name";
+        public const string HostNameHeader = "x-ms-functions-host-name";
+        public const string SiteTokenHeader = "x-ms-site-restricted-token";
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: Can you please define this one in ScriptConstants where we define some of the other headers? There are other places in the code where we use this header name - it'd be good to have them all use the same constant.

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        private void SendRequest<T>(ConcurrentQueue<T> activitiesToPublish, string publishPath)
+        {
+            var request = BuildRequest(HttpMethod.Post, publishPath, activitiesToPublish.ToArray());
+            HttpResponseMessage response = _httpClient.SendAsync(request).GetAwaiter().GetResult();
+            response.EnsureSuccessStatusCode();
+        }
+
+        public void InitializeRequestHeaders()
+        {
+            _hostName = _environment.GetEnvironmentVariable(EnvironmentSettingNames.AzureWebsiteHostName);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: Use HostNameProvider for this - it handles slot swaps correctly. See other usages.

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+        }
+
+        private void SendRequest<T>(ConcurrentQueue<T> activitiesToPublish, string publishPath)
+        {
+            var request = BuildRequest(HttpMethod.Post, publishPath, activitiesToPublish.ToArray());
+            HttpResponseMessage response = _httpClient.SendAsync(request).GetAwaiter().GetResult();
+            response.EnsureSuccessStatusCode();
+        }
+
+        public void InitializeRequestHeaders()
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: Recommend making this a private method, and renaming to Initialize, since it does more than initialize header values.

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+            _processMonitorTimer = new Timer(OnProcessMonitorTimer, null, TimeSpan.Zero, _memorySnapshotInterval);
+            _metricsPublisherTimer = new Timer(OnFunctionMetricsPublishTimer, null, TimeSpan.Zero, _metricPublishInterval);
+
+            _logTraceEvent(LogLevel.Information, string.Format("LinuxContainerMetricsPublisher: Starting metrics publisher for container : {0}", _containerName));
+        }
+
+        private void OnProcessMonitorTimer(object state)
+        {
+            _process.Refresh();
+            var commitSizeBytes = _process.WorkingSet64;
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: Need to run this by @paulbatum to make sure this is the right value to use.

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        {
+            try
+            {
+                if (currentActivities.Any())
+                {
+                    SendRequest(currentActivities, publishPath);
+                }
+                DrainActivities(currentActivities, activitiesToProcess);
+            }
+            catch (Exception ex)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: Can you use the IsFatal extension method here to avoid swallowing fatal exceptions?

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+            _logTraceEvent = logMetricsPublishEvent ?? throw new ArgumentNullException(nameof(logMetricsPublishEvent));
+            _memoryActivities = new BlockingCollection<MemoryActivity>(new ConcurrentQueue<MemoryActivity>(), boundedCapacity: _maxBufferSize);
+            _functionActivities = new BlockingCollection<FunctionActivity>(new ConcurrentQueue<FunctionActivity>(), boundedCapacity: _maxBufferSize);
+
+            _currentMemoryActivities = new ConcurrentQueue<MemoryActivity>();
+            _currentFunctionActivities = new ConcurrentQueue<FunctionActivity>();
+
+            var nodeIpAddress = _environment.GetEnvironmentVariable(EnvironmentSettingNames.NodeIpAddress);
+            _requestUri = string.Format(_requestUriFormat, nodeIpAddress);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: Some of these you're initializing before starting the publisher, some only after. Why?

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+            FunctionActivity activity = new FunctionActivity
+            {
+                FunctionName = functionName,
+                InvocationId = invocationId,
+                Concurrency = concurrency,
+                ExecutionStage = executionStage,
+                IsSucceeded = success,
+                ExecutionTimeSpanInMs = executionTimeSpan,
+                EventTimeStamp = eventTimeStamp,
+                Tenant = _environment.GetEnvironmentVariable(EnvironmentSettingNames.WebSiteStampDeploymentId)?.ToLowerInvariant()
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: Can this env var be cached along with the others you're already caching?

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+            };
+
+            if (!_functionActivities.TryAdd(activity))
+            {
+                _logTraceEvent(LogLevel.Warning, string.Format("LinuxContainerMetricsPublisher: Buffer full. Dropping current batch of function activities"));
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: After you declare an ILogger<LinuxContainerMetricsPublisher> ctor param for logging, it will log the Type name as the source, so you don't have to hard code in the trace event.

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+        {
+            _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
+            _standbyOptions = standbyOptions ?? throw new ArgumentNullException(nameof(standbyOptions));
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+            _logTraceEvent = logMetricsPublishEvent ?? throw new ArgumentNullException(nameof(logMetricsPublishEvent));
+            _memoryActivities = new BlockingCollection<MemoryActivity>(new ConcurrentQueue<MemoryActivity>(), boundedCapacity: _maxBufferSize);
+            _functionActivities = new BlockingCollection<FunctionActivity>(new ConcurrentQueue<FunctionActivity>(), boundedCapacity: _maxBufferSize);
+
+            _currentMemoryActivities = new ConcurrentQueue<MemoryActivity>();
+            _currentFunctionActivities = new ConcurrentQueue<FunctionActivity>();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: What's the difference between these pairs of collections? Can you please add comments to make their roles clear?

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+                {
+                    currentActivites.Enqueue(item);
+                }
+            }
+        }
+
+        internal void OnFunctionMetricsPublishTimer(object state)
+        {
+            PublishActivity(_currentMemoryActivities, _memoryActivities, PublishMemoryActivityPath);
+            PublishActivity(_currentFunctionActivities, _functionActivities, PublishFunctionActivityPath);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: I don't understand these 2 collections. Looks like on Add you're adding only to _functionActivities, but PublishActivity is only Sending _currentActivities?

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+                Tenant = _environment.GetEnvironmentVariable(EnvironmentSettingNames.WebSiteStampDeploymentId)?.ToLowerInvariant()
+            };
+
+            if (!_functionActivities.TryAdd(activity))
+            {
+                _logTraceEvent(LogLevel.Warning, string.Format("LinuxContainerMetricsPublisher: Buffer full. Dropping current batch of function activities"));
+                DrainActivities(_currentFunctionActivities, _functionActivities);
+            }
+
+            _logTraceEvent(LogLevel.Information, string.Format("LinuxContainerMetricsPublisher: Added function activity : {0} {1} {2} {3} {4} {5}", functionName, invocationId, concurrency, executionStage, success, executionTimeSpan));
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: Recommend using string interpolation for all these to make them easier to read/verify. E.g. $"Added function activity: {functionName}, ..."

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+            if (activitiesToProcess.Any())
+            {
+                while (activitiesToProcess.TryTake(out T item))
+                {
+                    currentActivites.Enqueue(item);
+                }
+            }
+        }
+
+        internal void OnFunctionMetricsPublishTimer(object state)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: Make this "async void" so the publish activities can be properly async, without the GetAwaiter().GetResult()s

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+        }
+
+        public void InitializeRequestHeaders()
+        {
+            _hostName = _environment.GetEnvironmentVariable(EnvironmentSettingNames.AzureWebsiteHostName);
+            _stampName = _environment.GetEnvironmentVariable(EnvironmentSettingNames.WebSiteHomeStampName);
+            _process = Process.GetCurrentProcess();
+        }
+
+        public void StartMetricsPublisher()
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: nit: just call it Start

File: src/WebJobs.Script.WebHost/Models/FunctionActivity.cs [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+        public string ExecutionId { get; set; }
+
+        public string ExecutionStage { get; set; }
+
+        public bool IsSucceeded { get; set; }
+
+        public long ExecutionTimeSpanInMs { get; set; }
+
+        public DateTime EventTimeStamp { get; set; }
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Models/FunctionActivity.cs]: Add a common base class to hold the common properties (EventTimeStamp, Tenant, etc.). Could call it ActivityBase

File: src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs [hunk 18]
```diff
@@ -85,7 +85,9 @@ public static void AddWebJobsScriptHost(this IServiceCollection services, IConfi
                 var environment = p.GetService<IEnvironment>();
                 if (environment.IsLinuxContainerEnvironment())
                 {
-                    return new LinuxContainerEventGenerator(environment);
+                    var standbyOptions = p.GetService<IOptionsMonitor<StandbyOptions>>();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs]: Should be able to undo these changes, once you move the publisher to DI

File: src/WebJobs.Script/Environment/EnvironmentSettingNames.cs [hunk 19]
```diff
@@ -64,6 +64,8 @@ public static class EnvironmentSettingNames
         public const string RunningInContainer = "DOTNET_RUNNING_IN_CONTAINER";
 
         public const string ExtensionBundleSourceUri = "FUNCTIONS_EXTENSIONBUNDLE_SOURCE_URI";
+
+        public const string NodeIpAddress = "Fabric_NodeIPOrFQDN";
```
Comment (mathewc) [inline src/WebJobs.Script/Environment/EnvironmentSettingNames.cs]: nit: LinuxNodeIpAddress?

File: src/WebJobs.Script.WebHost/Diagnostics/LinuxContainerEventGenerator.cs (hunk 1 above)
Comment (divyagandhii) [reply src/WebJobs.Script.WebHost/Diagnostics/LinuxContainerEventGenerator.cs]: I tried DI that but it leads to circular dependencies in DI which is why I am doing it this way

File: src/WebJobs.Script.WebHost/Diagnostics/LinuxContainerEventGenerator.cs (hunk 2 above)
Comment (divyagandhii) [reply src/WebJobs.Script.WebHost/Diagnostics/LinuxContainerEventGenerator.cs]: We had to have a workaround for diagnostics when we were rolling out on LCOW. We don't need this workaround anymore so I am removing it.

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs (hunk 9 above)
Comment (divyagandhii) [reply src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: The ones initialized after starting the publisher are the ones we get only after we are specialized. We cannot get those values when we create the publisher.

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs (hunk 7 above)
Comment (divyagandhii) [reply src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: We had checked this with the cosine team as well.

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs (hunk 6 above)
Comment (divyagandhii) [reply src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: The tests only validate that request headers are set correctly

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs (hunk 10 above)
Comment (divyagandhii) [reply src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: This setting is also only available after specialization. We will now cache it when we initialize.

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs (hunk 13 above)
Comment (divyagandhii) [reply src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: Yes _currentactivities is the current batch of activities we are trying to send. _functionactivities is the queue buffer where we store all activities until they are picked up by the current activities batch

File: src/WebJobs.Script.WebHost/Diagnostics/LinuxContainerEventGenerator.cs [hunk 20]
```diff
@@ -95,6 +99,11 @@ private void ConsoleWriter(string evt)
             }
         }
 
+        private void LogMetricsPublishEvent(LogLevel level, string message)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Diagnostics/LinuxContainerEventGenerator.cs]: This function doesn't appear to be used? This line and the unused commented out line above?

File: src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs [hunk 21]
```diff
@@ -26,8 +27,9 @@ public class MetricsEventManager : IDisposable
         private static string appName;
         private static string subscriptionId;
         private bool _disposed;
+        private IMetricsPublisher _metricsPublisher;
 
-        public MetricsEventManager(IEnvironment environment, IEventGenerator generator, int functionActivityFlushIntervalSeconds, int metricsFlushIntervalMS = DefaultFlushIntervalMS)
+        public MetricsEventManager(IEnvironment environment, IEventGenerator generator, int functionActivityFlushIntervalSeconds, IMetricsPublisher metricsPublisher = null, int metricsFlushIntervalMS = DefaultFlushIntervalMS)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs]: I think you're allowing null just so the test doesn't have to pass in a mock? Let's not take null and just pass in a mock in the tests, so it's clear that this param is required, just like generator, etc.

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+                ExecutionStage = executionStage,
+                IsSucceeded = success,
+                ExecutionTimeSpanInMs = executionTimeSpan,
+                EventTimeStamp = eventTimeStamp,
+                Tenant = _tenant
+            };
+
+            if (!_functionActivities.TryAdd(activity))
+            {
+                _logger.LogWarning("Buffer full.Dropping current batch of function activities");
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: nit: fix error message

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+                DrainActivities(_currentFunctionActivities, _functionActivities);
+            }
+
+            _logger.LogInformation($"Added function activity : {functionName} {invocationId} {concurrency} {executionStage} {success} {executionTimeSpan}");
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: These should all probably be Debug level, right? (logger.LogDebug).

File: src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs [hunk 24]
```diff
@@ -0,0 +1,235 @@
...
+        public void AddMemoryActivity(DateTime timeStampUtc, long data)
+        {
+            var memoryActivity = new MemoryActivity
+            {
+                CommitSizeInBytes = data,
+                EventTimeStamp = timeStampUtc,
+                Tenant = _tenant
+            };
+
+            if (!_memoryActivities.TryAdd(memoryActivity))
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: I think this code would be simpler if you just used a single collection for each activity type, not the dual collection with copy operations that you're doing now. Just like we do in FileWriter.cs - we just have a single buffer we're writing to, and when we flush we just snapshot the current contents for publish within a lock. Same problem as here.

Also in FileWriter, when we hit limit we don't have any drain/copy operations to perform - we just stop taking new entries. Why when you hit the limit here do you drop all the _previous_ logs rather than the new one coming in?

Comment (divyagandhii) [reply src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs]: Because it is possible something is wrong with the current batch of events we are trying to send. In those cases, we just want to skip the current batch and pick up the next one. The other scenario is that metrics publisher fails because the metering service is down for an extended period of time. In those cases as well, we should drop the older events rather than the new ones.
